pkg/rpc/request/types.go

```diff
@@ -35,43 +37,106 @@ type Raw struct {
 	ID        int           `json:"id"`
 }
 
-// In represents a standard JSON-RPC 2.0
-// request: http://www.jsonrpc.org/specification#request_object. It's used in
-// server to represent incoming queries.
+// In contains standard JSON-RPC 2.0 request and batch of
+// requests: http://www.jsonrpc.org/specification.
+// It's used in server to represent incoming queries.
 type In struct {
+	Request *Request
+	Batch   Batch
+}
+
+// Request represents a standard JSON-RPC 2.0
+// request: http://www.jsonrpc.org/specification#request_object.
+type Request struct {
 	JSONRPC   string          `json:"jsonrpc"`
 	Method    string          `json:"method"`
 	RawParams json.RawMessage `json:"params,omitempty"`
 	RawID     json.RawMessage `json:"id,omitempty"`
+	Error     error           `json:"-"`
 }
 
-// NewIn creates a new Request struct.
-func NewIn() *In {
-	return &In{
-		JSONRPC: JSONRPCVersion,
+// Batch represents a standard JSON-RPC 2.0
+// batch: https://www.jsonrpc.org/specification#batch.
+type Batch []*Request
+
+// MarshalJSON implements json.Marshaler interface
+func (i In) MarshalJSON() ([]byte, error) {
+	if i.Request != nil {
+		return json.Marshal(i.Request)
 	}
+	return json.Marshal(i.Batch)
+}
+
+// UnmarshalJSON implements json.Unmarshaler interface.
+func (i *In) UnmarshalJSON(data []byte) error {
+	var (
+		req   *Request
+		batch Batch
+	)
+	switch data[0] {
+	case '[':
+		decoder := json.NewDecoder(bytes.NewReader(data))
+		decoder.Token() // read `[`
+		for decoder.More() {
```

Comment:
BTW, there has to be some limit on how many requests could be in a single batch.

Reply:
JSON-RPC specification does not set such limit. I implemented restriction = 100 requests per batch, but we can easily changed that to a new number.

Reply:
Yeah, the spec doesn't have it, but practically there has to be some restriction.